src/ripple/app/misc/impl/ValidatorSite.cpp

```diff
@@ -252,6 +269,32 @@ ValidatorSite::makeRequest (
 
     work_ = sp;
     sp->run ();
+    // start a timer for the request, which shouldn't take more
+    // than requestTimeout_ to complete
+    std::unique_lock <std::mutex> lock_state{state_mutex_};
+    timer_.expires_at ( clock_type::now() + requestTimeout_);
+    timer_.async_wait ( std::bind (
+        &ValidatorSite::onRequestTimeout, this, siteIdx, std::placeholders::_1));
+}
+
+void
+ValidatorSite::onRequestTimeout (
+    std::size_t siteIdx,
+    error_code const& ec)
+{
+    if (ec)
+        return;
+
+    {
+        std::unique_lock <std::mutex> lock_site{sites_mutex_};
```

Comment:
While you are here... it seems the caught exception at line 327, `onSiteFetch` is called while already locked on `sites_mutex`. I believe that is UB.

It would also be nice to add a documenting param to `setTimer`  to make it clear that it is always called with a lock on `state_mutex_`

Reply:
yep - good point. I'll move the lock inside the try and that should fix it. I'll also remove that assert since we don't have a state lock to protect that access (strictly..perhaps this could have just been an atomic). I'll add the lock param to setTimer too.